lib/imports/api/authority/server/authority.js

```diff
@@ -88,25 +96,72 @@ export class Authority {
       clientIds: this.stepClientIds.slice(version)
     };
   }
-  // /**
-  //  * Store a snapshot of the current document in the documentsCollection
-  //  */
-  // storeSnapshot() {
-  //   const doc = this.pm.doc;
-  //   const newSnapshotJSON = doc.toJSON();
-  //   const newSnapshotNumber = ProseMeteorManager.numberOfSnapshots() + 1;
-  //   this.documentsCollection.insert({
-  //     docId: this.docId,
-  //     docJSON: newSnapshotJSON,
-  //     timestamp: Date.now(),
-  //     number: newSnapshotNumber
-  //   });
-  // }
+
+  /**
+  * Returns total number of snapshots in db for this Authority's document.
+  * @returns {Number}
+  */
+  numberOfSnapshots(){
+    return this.documentsColl.find({ docId: this.docId }).count();
+  }
+
+  /**
+  * Starts the timer interval to store doc snapshots to db.
+  */
+  startSnapshotInteval() {
+    this.snapshotIntervalTimer = Meteor.setInterval(() => {
+      // only store snapshot if changes have occurred since last snapshot
+      let latestSnapshotVersion = this.latestSnapshot.version;
+      let version = this.version();
+      if (version > latestSnapshotVersion) {
+        console.log('Authority storing snapshot for doc "' + this.docId + '", version ' + version);
+        this.storeSnapshot();
+      }
+    }, this.snapshotIntervalMs);
+  }
+
+  /**
+  * Clears the snapshot interval.
+  */
+  stopSnapshotInterval() {
+    Meteor.clearInterval(this.snapshotIntervalTimer);
+  }
+
+  /**
+  * Returns the version number of this Authority's document.
+  * @returns {Number}
+  */
+  version() {
+    let version;
+    if (this.numberOfSnapshots() === 0) {
+      version = this.steps.length;
+    } else {
+      // if there's a previous snapshot, current doc version is last snapshot's version + the number of steps since then
+      let latestSnapshotVersion = this.latestSnapshot.version;
+      version = latestSnapshotVersion + this.stepsSince({ version: latestSnapshotVersion }).steps.length;
+    }
+    return version;
+  }
+
```

Comment:
Since the doc can be loaded from a snapshot, just using `this.steps.length` doesn't work anymore since our previous implementation assumed all steps are in memory. If I open a doc and it loads it's last snapshot, which was at version 1200, and then I make 5 edits, `this.steps.length=5` but the actual doc version is 1205 (the latest snapshot version 1200 + the 5 edits since then). I actually overlooked that part in `latestDocState`, I'll have to change it to use `this.version()`